cterasdk/edge/firmware.py

```diff
@@ -0,0 +1,55 @@
+from ..lib import FileSystem
+from ..exception import CTERAException
+from .base_command import BaseCommand
+
+
+class UploadTaskStatus():
+    IN_PROGRESS = 0
+    COMPLETE = 1
+    FAIL = -1
+
+
+class Firmware(BaseCommand):
+    """ Gateway Firmware upgrade API """
+
+    def __init__(self, gateway):
+        super().__init__(gateway)
+        self._filesystem = FileSystem.instance()
+
+    def upgrade(self, file_path, reboot=True, wait_for_reboot=True):
+        """
+        Upgrade the Filer F/W with the provided file
+
+        :param str file_path: Path to the local file to upload
+        :param bool,optional reboot: Perform reboot after uploading the new F/W, defaults to True
+        :param bool,optional wait_for_reboot: Wait for reboot to complete (if reboot is performed), defaults to True
+        """
+        upload_task_info = self._upload_firmware(file_path)
+        if upload_task_info.rc != 0:
+            raise CTERAException(message='Failed to upload the new firmware', path=file_path)
+        self._wait_for_completion(upload_task_info.taskPointer)
+        if reboot:
+            self._gateway.power.reboot(wait=wait_for_reboot)
+
+    def _upload_firmware(self, file_path):
+        file_info = self._filesystem.get_local_file_info(file_path)
+        with open(file_path, 'rb') as fd:
+            return self._gateway.upload(
+                'proc/firmware',
+                dict(
+                    name='upload',
+                    firmware=(file_info['name'], fd, file_info['mimetype'][0])
+                )
+            )
+
+    def _wait_for_completion(self, task_pointer):
```

Comment:
There's an edge/taskmgr.py module used to monitor the progress of background tasks. Can you use that instead?

Reply:
I wanted to. But, the mechanism is handled differently -
1. The object returned from the upload command is different
2. Plus, it contains the entire URL (not a bgtask with UID)
3. The object returned from the Get status URL is different
4. All of this is specific to the firmware upgrade sequence (I looked at the G/W code)
As a result, this required different code that is not generic and hence will not be used anywhere else. So, I don't see the advantage of putting it anywhere outside of the scope of this class

Reply:
I see. Ok. Btw, there's also a ``edge/tracker.py`` for scenario's where you monitor a specific schema attribute (not a BgTask). But perhaps it's not applicable in this scenario either.